[V1][Metrics] Handle preemptions

markmc opened the pull request:
Part of #10582 

Add a core engine `PREEMPTED` event.

Add the `num_preemptions_total` counter from v0.

Also, make preemptions reset the scheduled and first token timestamps resulting in:

```
  << queued timestamp >>
    [ queue interval ]
  << scheduled timestamp (FIRST) >>
    [ prefill interval ]
  << new token timestamp (FIRST) >>
    [ inter-token interval ]
  << new token timestamp >>
    [ inter-token interval ]
      |
      |	(possible preemptions)
      | << scheduled timestamp >>
      | << preempted timestamp >>
      | << scheduled timestamp >>
      | << new token timestamp >>
      | << preempted timestamp >>
      v
  << new token timestamp >>
    [ decode interval (relative to first token time)
    [ inference interval (relative to first scheduled time)
  << new token timestamp (FINISHED) >>
```

markmc committed:
[V1][Metrics] Handle preemptions

Add a core engine PREEMPTED event.

Add the num_preemptions_total counter from v0.

Also, make preemptions reset the scheduled and first token timestamps
resulting in:

```
  << queued timestamp >>
    [ queue interval ]
      |
      |	(possible preemptions)
      | << scheduled timestamp >>
      | << preempted timestamp >>
      | << scheduled timestamp >>
      | << new token timestamp (FIRST) >>
      | << preempted timestamp >>
      v
  << scheduled timestamp >>
    [ prefill interval ]
  << new token timestamp (FIRST) >>
    [ inter-token interval ]
  << new token timestamp >>
    [ decode interval (relative to most recent first token time)
    [ inference interval (relative to most recent scheduled time)
  << new token timestamp (FINISHED) >>
```

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc commented:
pre-commit failure is a `yapf` failure that doesn't happen for me locally:

```
  File "/home/runner/.cache/pre-commit/repom9gt4aao/py_env-python3.12/lib/python3.12/site-packages/yapf_third_party/_ylib2to3/pygram.py", line 39, in <module>
    pattern_grammar = driver.load_grammar(_PATTERN_GRAMMAR_FILE)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/.cache/pre-commit/repom9gt4aao/py_env-python3.12/lib/python3.12/site-packages/yapf_third_party/_ylib2to3/pgen2/driver.py", line 248, in load_grammar
    g.load(gp)
  File "/home/runner/.cache/pre-commit/repom9gt4aao/py_env-python3.12/lib/python3.12/site-packages/yapf_third_party/_ylib2to3/pgen2/grammar.py", line 128, in load
    d = pickle.load(f)
        ^^^^^^^^^^^^^^
EOFError: Ran out of input
```

robertgshaw2-redhat reviewed (changes requested):
From the POV of a client, "preemptions" create bumps in ITL (inter token latency) aka TPOT (time per output token).

You can see this clearly when considering streaming:
- The user sends a prompt like `Hello my name is Robert and I`
- The model streams back `like` -> `to` -> `work` -> `on` -> `vllm`

If a request is preempted after the word `to`, we will evict the request and re-run once we have enough KV cache memory with the prompt as "Hello my name is Robert and I like to" (so the generated tokens are added to the prompt when recomputing --- this makes the recomputation happen much more quickly --- and prefix caching also helps to make this happen fast).

From the POV of the user, this means that the ITL of the word "work" will be elevated (since all the time associated with being preempted, waiting, and then recomputing will happen before the word is emitted). The TTFT is not impacted since we already streamed the word `like`. As a result, I think that we should make the metrics coming out of vLLM match this structure
- When a `PREEMPTION` occurs, this should manifest as a higher `inter token interval` for that token
- When a `PREEMPTION` occurs, this should manifest as higher `decode interval` for that request

robertgshaw2-redhat commented:
So I think that it should look like this:

```bash
<< queued timestamp >>  # unique per request and frozen after being set
    [ queue interval ]
<< scheduled timestamp >> # unique per request and frozen after being set
    [ prefill interval ]
<< new token timestamp (FIRST) >> # unique per request and frozen after being set
    [ inter-token interval ]
<< new token timestamp >>
<< preempted timestamp >>
    | request is in the preempted queue
    | request is re-scheduled
    | recompute up to the current token
    [ inter-token interval ]
<< new token timestamp >>
    [ inter-token interval ]
<< new token timestamp >>
<< preempted timestamp >>
    | request is in the preempted queue
    | request is re-scheduled
    | recompute up to the current token
    [ inter-token interval ]
  << new token timestamp >>
    [ decode interval (relative to scheduled)
    [ inference interval (relative to scheduled)
  << new token timestamp (FINISHED) >>
```

robertgshaw2-redhat commented:
One other case we should make sure we cover is --- what happens if the request is preempted during prefill? 

E.g. if we have a prompt of length 100k and we get preempted half way through. In this case, the preemption/recomputation time should be allocated to the prefill phase and the TTFT phase

Im not sure if this is possible to happen or whether there is some invariant that we always have enough KV cache for the prompt to be processed. Worth asking Woosuk or Cody

markmc commented:
I drafted some thoughts, and found it super difficult to clarify this in words or ASCII art so let's try this ... does this match your thinking?

![vLLM Interval Metrics - Frame 1](https://github.com/user-attachments/assets/af8594e7-34bd-4e66-978e-1f897bd46100)
![vLLM Interval Metrics - Frame 2](https://github.com/user-attachments/assets/9b2fb370-836b-4b8a-8b16-d6860844a3ac)
![vLLM Interval Metrics - Frame 3](https://github.com/user-attachments/assets/be82c60d-6f40-46d9-bf33-c48a855b577b)

markmc commented:
In the "preempted prefill" case, I had imagined the queued interval to be up until the final `SCHEDULED` event ... nothing useful happened with the request, its waiting to be prioritized for resources? Not a big deal, I guess - most important that `TTFT ~= queued_interval + prefill_interval`

If we're aligned on the diagrams above, I think the code change is simply to not reset `scheduled_ts` or `first_token_ts` once they've been set?

robertgshaw2-redhat commented:
I am in alignment with the charts above! Thanks for drawing it out!

markmc committed:
[V1][Metrics] Change where preemption time is accounted

Don't include prefill preemption time in the queued interval.

Don't reset first token on preemption - already decoded tokens
are retained and reused.

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

robertgshaw2-redhat reviewed (approved):
This looks great and is very robust. Thanks!
